### ISSUE TITLE
Unable to pickle redis-py instance (_thread.lock)

### ISSUE THREAD

**Issue by kjanko** [andymccurdy/redis-py#1029]
So I'm defining multiple workers in the background, and each worker has a connection instance created on startup:

```
class Worker(Process):
    _db = None

    def __init__(self):
        super(Worker, self).__init__()
        self._db = redis.Redis(host="1.2.3.4", port=1234, db=0)
```

However, whenever I try to start an instance of this worker, I get the following error message:

    

> TypeError: can't pickle _thread.lock objects

So I'm guessing that this implementation uses a lock somewhere. What is the workaround for this issue?

**Comment by kjanko** [comment]
The issue seems to persist only if I try to define the connection in the constructor. When defined inside the run() method, It works fine. What's the explanation behind this behavior?

**Comment by Grokzen** [comment]
@kjanko There is 2 main options of parralell execution in python (threads and processes). Threads live inside the same process as the python you start. But processes spin up child processes/copies to the main process to enable parralell execution. When dealing with threads you need to have the lock in place in your code to protect against problems with multiple parties claiming the same resources and cause deadlocks etc. Processes however have the issue that when you make a clone of yourself, you have to send down a memory representation of the data from the parent to all child processes so they can independently execute without the need for locks between the processes. Quoted from a blog post
```
As mentioned by Matthew Rockling in his article Parallelism and Serialization, processes send information much like the teleporters in Star Trek: a package is converted to something understandable by the teleporter, sent, and transformed back into its package form. This process is called serialization, and Python uses the Pickle library to handle translating the object back and forth.
```
So, basically with your code, what you are doing is to create the redis object in the parent process and then it wants to serialize it down with pickle to send it to it's child processes. But since a lock is built into this code base, it can't be done.

The only solution you have at your disposal to work with processes is to do what you specified in your first comment that you must define it inside the run() method as when you get into the scope of the run() method then you are already inside the child process and it can spawn the redis object inside that scope and it will let you use threads inside it if you like.

This issue can be closed imo, with the `run()` implementation is the correct solution here.

**Comment by andymccurdy** [comment]
Thanks @Grokzen. I agree that this is an acceptable solution.

**Comment by coleifer** [comment]
>The only solution you have at your disposal to work with processes is to do what you specified in your first comment that you must define it inside the run() method as when you get into the scope of the run() method then you are already inside the child process and it can spawn the redis object inside that scope and it will let you use threads inside it if you like.

This isn't quite correct. If you use the `fork` method on an initialized redis connection pool, everything behaves sanely (and I seem to recall redis-py also has a fork-checker built-in). It seems at least reasonable to expect it to work with `spawn` method as well. Possibly redis-py can lazily instantiate the `threading.Lock` when creating the `ConnectionPool` here: https://github.com/redis/redis-py/blob/master/redis/connection.py#L1372

My sense is that just instantiating a connectionpool without checking-out any connections should be OK, especially when using the `spawn` multiprocessing method.
